### ISSUE TITLE
exception in ParameterCompatibleWithTypeConstraint when parameter type is interface and constraint type is class

### ISSUE THREAD

**Issue by polivbr** [autofac/Autofac#1009]
Hello, 

I have found a bug introduced in 4.9.3, where an ArgumentNullException is thrown within ParameterCompatibleWithTypeConstraint when the parameter type is an interface, and the constraint type is a class. This did not occur in 4.9.2.

The exception is thrown from the call to "baseType.GetTypeInfo()", as the baseType is null because the parameter is an interface and thus has no base class.

Below is a simple program that demonstrates the issue.

```
using Autofac;

namespace AutofacBug
{
    class Program
    {
        static void Main(string[] args)
        {
            using (var container = BuildContainer())
            {
                var handler = container.Resolve<IHandler<IStuff>>();
            }
        }

        static IContainer BuildContainer()
        {
            var builder = new ContainerBuilder();

            builder
                .RegisterGeneric(typeof(GenericHandler<>))
                .AsImplementedInterfaces();

            builder
                .RegisterType<StuffHandler>()
                .AsImplementedInterfaces();

            return builder.Build();
        }

        interface IHandler<T>
        {
        }

        class StuffHandler : IHandler<IStuff>
        {
        }

        class GenericHandler<T> : IHandler<T> where T: GenericBase
        {
        }

        interface IStuff
        {
        }

        class GenericBase
        {
        }
    }
}
```

**Comment by nblumhardt** [comment]
Duplicate of #1006 I think. Thanks!

**Comment by bpolivka** [comment]
As far as I can tell, #1006 is actually a different bug in the same method. Doesn’t look like the fix addresses the same issue either.

**Comment by nblumhardt** [comment]
Thanks for looking into it, sorry I missed that detail. 👍 

If anyone has a chance to convert the above into a unit test and PR, please jump on in :-)

**Comment by smartcodinghub** [comment]
I have this exact issue. I downgrade to 4.8.1, but reading that the 4.9.2 is working I'll test again so we have at least a workaround.

**Comment by alistairjevans** [comment]
I fixed #1006, since this is in the same area I'll sort out a PR for the test and the fix.

**Comment by alexmg** [comment]
Fix available in v4.9.4 on NuGet
